src/dataflow/src/render/debezium.rs

```diff
@@ -118,6 +121,265 @@ where
+                                        *data_cap.time(),
+                                        1,
+                                    ));
+                                    let _ = data_buffer.pop_front();
+                                    continue;
+                                }
+                                None => {
+                                    let _ = data_buffer.pop_front();
+                                    continue;
+                                }
+                            };
+
+                            let tx_id = value.iter().nth(transaction_idx).unwrap().unwrap_list().iter().nth(tx_metadata_description.data_transaction_id_idx).unwrap().unwrap_str().to_owned();
+
+                            let tx_time: Timestamp = match tx_mapping.get(&tx_id) {
+                                Some(time) => *time,
+                                None => break,
+                            };
+
+                            // We're committed to processing it if we have a tx_time
+                            let _ = data_buffer.pop_front();
+
+                            let partition_dedup = dedup_state
+                                .entry(result.partition.clone())
+                                .or_insert_with(|| {
+                                    DebeziumDeduplicationState::new(envelope.clone())
+                                });
+                            let should_use = match partition_dedup {
+                                Some(ref mut s) => {
+                                    let res = s.should_use_record(
+                                        key,
+                                        &value,
+                                        result.position,
+                                        result.upstream_time_millis,
+                                        &debug_name,
+                                    );
+                                    match res {
+                                        Ok(b) => b,
+                                        Err(err) => {
+                                            // We could theoretically use tx_cap_map.get(tx_id) here but for sake of
+                                            // consistency, we output all errors at the data_cap time.
+                                            output.session(&data_cap).give((
+                                                Err(err),
+                                                *data_cap.time(),
+                                                1,
+                                            ));
+                                            continue;
+                                        }
+                                    }
+                                }
+                                None => true,
+                            };
+
+                            if should_use {
+                                let tx_cap_entry = match tx_cap_map.entry(tx_id.clone()) {
```

Comment:
this looks like a `tx_cap_map.get(&tx_id).expect("Must have cap if using record")`

Reply:
I used the Entry API here to make the removal below more ergonomic -- so we only panic in one spot.  I was attempting to minimize explicit panics -- especially given how many we need to have given the data structures we're using to track things.  I think it's easier to reason about correctness this way.  What do you think?

Reply:
I missed the removal further down, sorry. I retract my suggestion, that looks fine